### PR TITLE
[master][KOGITO-1597] Refactor Cucumber tests to get rid of KogitoAppDeployment

### PR DESCRIPTION
**Pull request by sutaakar**
Signed-off-by: Karel Suta <ksuta@redhat.com>

https://issues.redhat.com/browse/KOGITO-1597

Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.MD#sending-a-pull-request)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [ ] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

### CODE REVIEW

File: pkg/apis/app/v1alpha1/kogitoapp_types.go [hunk 1]
```diff
@@ -132,6 +132,20 @@ type KogitoAppBuildObject struct {
 	MavenMirrorURL string `json:"mavenMirrorURL,omitempty"`
 }
 
+// AddEnvironmentVariable adds new environment variable to build environment variables
+func (k *KogitoAppBuildObject) AddEnvironmentVariable(name, value string) {
+	if k == nil {
```
Comment (Sgitario) [inline pkg/apis/app/v1alpha1/kogitoapp_types.go:137]: Does it make sense this check? I mean using this method by:

```go
k.AddEnvironmentVariable where k is null
```

File: pkg/apis/app/v1alpha1/kogitoservices_types.go [hunk 2]
```diff
@@ -157,3 +157,17 @@ func (k *KogitoServiceSpec) GetResources() corev1.ResourceRequirements { return
 func (k *KogitoServiceSpec) SetResources(resources corev1.ResourceRequirements) {
 	k.Resources = resources
 }
+
+// AddEnvironmentVariable adds new environment variable to service environment variables
+func (k *KogitoServiceSpec) AddEnvironmentVariable(name, value string) {
+	if k == nil {
```
Comment (Sgitario) [inline pkg/apis/app/v1alpha1/kogitoservices_types.go:163]: Same question as above.

File: pkg/apis/app/v1alpha1/kogitoapp_types.go (hunk 1 above)
Comment (sutaakar) [reply pkg/apis/app/v1alpha1/kogitoapp_types.go:137]: Good question, I have put it there as it is done i other methods.
Will try if it is usable.

Comment (sutaakar) [reply pkg/apis/app/v1alpha1/kogitoapp_types.go:137]: Actually, the nil check has sense. For example this is a "valid" code:
```
	var spec *KogitoServiceSpec
	spec.AddEnvironmentVariable("qq", "ww")
```
Function `AddEnvironmentVariable` is called with `spec = nil`

**Comment by sutaakar** [comment]
Cucumber tests passed (apart from Jobs service which is broken now)
Ready for review.

File: pkg/apis/app/v1alpha1/kogitoservices_types.go [hunk 3]
```diff
@@ -157,3 +157,17 @@ func (k *KogitoServiceSpec) GetResources() corev1.ResourceRequirements { return
 func (k *KogitoServiceSpec) SetResources(resources corev1.ResourceRequirements) {
 	k.Resources = resources
```
Comment (Sgitario) [inline pkg/apis/app/v1alpha1/kogitoservices_types.go:158]: I think we should make a decision about this. Either we check null pointers everywhere, or nowhere or we change the approach about adding functions vs methods.
What do you think?

Comment (sutaakar) [reply pkg/apis/app/v1alpha1/kogitoservices_types.go:158]: I think we can use this approach:
Preferred way of declaring such methods is to declare them on struct, not pointer.
If method is declared on pointer then it should handle case when pointer is nil.

@ricardozanini @radtriste  What do you think about it?

Comment (radtriste) [reply pkg/apis/app/v1alpha1/kogitoservices_types.go:158]: good question. for me it should be checked before the method is called and not when it is called...

File: test/framework/kogitoapp.go [hunk 4]
```diff
@@ -34,144 +30,92 @@ import (
...
-	Labels             map[string]string
-}
-
 // DeployService deploy a Kogito service
-func DeployService(namespace string, installerType InstallerType, kogitoAppDeployment KogitoAppDeployment) error {
-	GetLogger(namespace).Infof("%s deploy %s example %s with name %s, native %v, persistence %v, events %v and labels %v", installerType, kogitoAppDeployment.Runtime, kogitoAppDeployment.ContextDir, kogitoAppDeployment.AppName, kogitoAppDeployment.Native, kogitoAppDeployment.Persistence, kogitoAppDeployment.Events, kogitoAppDeployment.Labels)
+func DeployService(namespace string, installerType InstallerType, kogitoApp *v1alpha1.KogitoApp) error {
+	GetLogger(namespace).Infof("%s deploy %s example %s with name %s, native %v, persistence %v, events %v and labels %v", installerType, kogitoApp.Spec.Runtime, kogitoApp.Spec.Build.GitSource.ContextDir, kogitoApp.Name, kogitoApp.Spec.Build.Native, kogitoApp.Spec.EnablePersistence, kogitoApp.Spec.EnableEvents, kogitoApp.Spec.Service.Labels)
+
+	setupBuildImageStreams(kogitoApp)
```
Comment (radtriste) [inline test/framework/kogitoapp.go:40]: should be called in `crDeployService` or in `cliDeployService` you should check against this kogitoApp and not the config

File: test/steps/kogitoapp.go [hunk 5]
```diff
@@ -173,3 +134,49 @@ func (data *Data) scaleKogitoApplicationToPodsWithinMinutes(name string, nbPods,
...
+	kogitoApp.Spec.Build.GitSource.URI = config.GetExamplesRepositoryURI()
+	kogitoApp.Spec.Build.GitSource.ContextDir = contextDir
+
+	if ref := config.GetExamplesRepositoryRef(); len(ref) > 0 {
+		kogitoApp.Spec.Build.GitSource.Reference = ref
+	}
+
+	// Add namespace for service discovery
+	// Can be removed once https://issues.redhat.com/browse/KOGITO-675 is done
+	kogitoApp.Spec.KogitoServiceSpec.AddEnvironmentVariable("NAMESPACE", namespace)
```
Comment (radtriste) [inline test/steps/kogitoapp.go:153]: don't we want to put that into the `framework.GetKogitoAppStub` ?
I mean it impacts all kogitoapp as soon as they use the service discovery... but we don't need to make the distinction

File: test/framework/kogitoapp.go [hunk 6]
```diff
@@ -34,144 +30,92 @@ import (
...
+
+	if kogitoApp.Spec.Build.Native {
 		cmd = append(cmd, "--native")
 	}
 
-	// Add namespace for service discovery
-	// Can be removed once https://issues.redhat.com/browse/KOGITO-675 is done
-	cmd = append(cmd, "-e", fmt.Sprintf("NAMESPACE=%s", namespace))
-
 	if mavenMirrorURL := config.GetMavenMirrorURL(); len(mavenMirrorURL) > 0 {
```
Comment (radtriste) [inline test/framework/kogitoapp.go:80]: check against kogitoApp CR

File: test/framework/kogitoapp.go [hunk 7]
```diff
@@ -34,144 +30,92 @@ import (
...
-	cmd = append(cmd, "--image-version", config.GetBuildImageVersion())
+	if buildImageVersion := config.GetBuildImageVersion(); len(buildImageVersion) > 0 {
+		cmd = append(cmd, "--image-version", buildImageVersion)
+	}
+	if buildS2iImageStreamTag := config.GetBuildS2IImageStreamTag(); len(buildS2iImageStreamTag) > 0 {
+		cmd = append(cmd, "--image-s2i", buildS2iImageStreamTag)
+	}
+	if buildRuntimeImageStreamTag := config.GetBuildRuntimeImageStreamTag(); len(buildRuntimeImageStreamTag) > 0 {
+		cmd = append(cmd, "--image-runtime", buildRuntimeImageStreamTag)
+	}
```
Comment (radtriste) [inline test/framework/kogitoapp.go:112]: check against kogitoApp

File: test/framework/kogitoapp.go [hunk 8]
```diff
@@ -187,14 +131,13 @@ func DeployServiceFromExampleFile(namespace, exampleFile string) error {
...
-	kogitoApp := getKogitoAppStub(namespace, "name-should-be overwritten-from-yaml", nil)
+	kogitoApp := GetKogitoAppStub(namespace, "name-should-be overwritten-from-yaml")
 
 	// Apply content from yaml file
 	if err := yaml.NewYAMLOrJSONDecoder(strings.NewReader(yamlContent), len([]byte(yamlContent))).Decode(kogitoApp); err != nil {
 		return fmt.Errorf("Error while unmarshalling file: %v ", err)
 	}
 
-	// Apply custom image configuration
 	setupBuildImageStreams(kogitoApp)
```
Comment (radtriste) [inline test/framework/kogitoapp.go:141]: shouldn't we put that in `GetKogitoAppStub` in fact ?

Comment (sutaakar) [reply test/framework/kogitoapp.go:141]: The point is that some yaml examples can declare image versions.

What should we do in such case? Test with images declared in yaml or test with images declared in test config?

**Comment by sutaakar** [comment]
@radtriste setupBuildImageStreams and namespace declaration moved to GetKogitoAppStub, all values in cliDeployService are taken from KogitoApp.

File: test/framework/kogitoapp.go (hunk 8 above)
Comment (radtriste) [reply test/framework/kogitoapp.go:140]: Ok I see your point. yeah should be let like that then :)

**Comment by radtriste** [comment]
@sutaakar ahah about `setupBuildImageStreams` I was just thinking about letting it like it was :S 

that will impact (for now) the travel agency and it has been moved to kogito-examples (or still in progress but soon enough) and will then follow the different versions of Kogito for now.
so better to override with the test version.

**Comment by sutaakar** [comment]
@Sgitario nil checks removed

**Pull request merged**